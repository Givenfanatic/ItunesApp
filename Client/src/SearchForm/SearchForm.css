.search-form {
    display: flex;
    flex-direction: column;
    align-items: center;
  }
  
  .result-container {
    height: 600px;
    width: 100%;
    overflow-y: scroll;
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
    grid-gap: 20px;
    margin-bottom: 5px;
  }

  .result-container h2 {
    margin-bottom: 10px;
    text-align: center;
    display: flex;
  }

  .result-item {
    background-color: #f7f7f7;
    border: 1px solid #e6e6e6;
    padding: 20px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }
  
  .result-item h3 {
    margin-top: 0;
  }
  
  .result-item p {
    margin-bottom: 0;
  }
  
  .result-item button {
    margin-top: 10px;
    align-self: flex;
  }  

  .form-control{
    margin-bottom: 7px;
  }