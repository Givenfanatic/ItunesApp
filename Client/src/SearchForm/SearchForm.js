import React, { useState } from 'react';
import { Form, Button } from 'react-bootstrap';
import './SearchForm.css';
import Alert from 'react-bootstrap/Alert';

function SearchForm() {
    //our state to set the search term
    const [searchTerm, setSearchTerm] = useState('');
    //our state to set the media type, initialised as all
    const [mediaType, setMediaType] = useState('all');
    //use states to set our results that we get from the back end
    const [results, setResults] = useState([]);
    //use state to set our favorites as the user adds them
    const [favorites, setFavorites] = useState([]);
    //state to display our success Alerts in JSX
    const [showSuccessAlert, setShowSuccessAlert] = useState(false);
    //state to display our error Alerts in JSX
    const [showErrorAlert, setShowErrorAlert] = useState(false);
    //this will let the user know if they have already added this favorite
    const [showAlreadyExists, setAlreadyExists] = useState(false);

    //function that will get the data to our back-end which will in turn go get the data from the itunes api and then return it
    function handleSubmit(event) {
        event.preventDefault();

        const searchUrl = `api/search?term=${searchTerm}&media=${mediaType}`;

        fetch(searchUrl)
            .then(response => response.json())
            .then(data => {
                //set our results
                setResults(data.results);
            })
            .catch(error => {
                console.error(error);
            });
    }

    //call to our backend to add the users favorites
    function addToFavorites(result) {
        const url = `api/favorites`;

        fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(result)
        })
            .then(response => {
                if (response.status === 200) {
                    setFavorites([...favorites, result]); //the spread operator (...) is used to create a new array that includes all the elements from the favorites array and adds the result at the end.
                    setShowSuccessAlert(true); // set the value of showSuccessAlert to true to display our success Alert
                    setTimeout(() => {
                        setShowSuccessAlert(false);
                    }, 1000); // hide the alert after 3 seconds
                } else if (response.status === 400) {
                    setShowErrorAlert(true); // set the value of showErrorAlert to true to show our error alert
                    setAlreadyExists(true); // will let the user know that the favorite already exists
                    setTimeout(() => {
                        setShowErrorAlert(false);
                        setAlreadyExists(false);
                    }, 1000); // hides the alerts after 3 seconds
                }
            })
            .catch(error => {
                console.error(error);
                setShowErrorAlert(true); // set the value of showErrorAlert to true
                setTimeout(() => {
                    setShowErrorAlert(false);
                }, 1000); // hide the alert after 3 seconds
            });
    }



    return (
        <div className="search-form">
            <Form onSubmit={handleSubmit}>
                <Form.Group controlId="formSearchTerm">
                    <Form.Label>Search Term</Form.Label>
                    <Form.Control type="text" value={searchTerm} onChange={event => setSearchTerm(event.target.value)} />
                </Form.Group>

                <Form.Group controlId="formMediaType">
                    <Form.Label>Media Type</Form.Label>
                    <Form.Control as="select" value={mediaType} onChange={event => setMediaType(event.target.value)}>
                        <option value="all">All</option>
                        <option value="movie">Movie</option>
                        <option value="podcast">Podcast</option>
                        <option value="music">Music</option>
                        <option value="audiobook">Audiobook</option>
                        <option value="shortFilm">Short Film</option>
                        <option value="tvShow">TV Show</option>
                        <option value="software">Software</option>
                        <option value="ebook">Ebook</option>
                    </Form.Control>
                </Form.Group>

                <Button variant="primary" type="submit" style={{ marginBottom: '10px' }}>Search</Button>
            </Form>

            {results.length > 0 ? <h2 style={{ textAlign: 'center' }}>Search Results:</h2> : null}
            {showSuccessAlert && (
                <Alert variant="success" onClose={() => setShowSuccessAlert(false)} dismissible>
                    Favorite successfully added!
                </Alert>
            )}
            {showErrorAlert && (
                <Alert variant="danger" onClose={() => setShowErrorAlert(false)} dismissible>
                    {showAlreadyExists ? 'Favorite already added' : 'Favorite could not be added!'}
                </Alert>
            )}
            <div className="result-container">
                {results.map(result => (
                    <div className="result-item" key={result.trackId}>
                        <img src={result.artworkUrl100} alt="Album artwork" />
                        <h3>{result.trackName}</h3>
                        <p>{result.artistName}</p>
                        <Button variant="outline-secondary" onClick={() => addToFavorites(result)}>Add to Favorites</Button>
                    </div>
                ))}
            </div>
        </div>
    );
}

export default SearchForm;
