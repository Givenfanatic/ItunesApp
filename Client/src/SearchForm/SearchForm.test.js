import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SearchForm from './SearchForm';

//test the UI of the search form
describe('SearchForm', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(<SearchForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
