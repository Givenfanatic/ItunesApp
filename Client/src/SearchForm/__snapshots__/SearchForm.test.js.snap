// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="search-form"
  >
    <form
      class=""
    >
      <div>
        <label
          class="form-label"
          for="formSearchTerm"
        >
          Search Term
        </label>
        <input
          class="form-control"
          id="formSearchTerm"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="form-label"
          for="formMediaType"
        >
          Media Type
        </label>
        <select
          class="form-control"
          id="formMediaType"
        >
          <option
            value="all"
          >
            All
          </option>
          <option
            value="movie"
          >
            Movie
          </option>
          <option
            value="podcast"
          >
            Podcast
          </option>
          <option
            value="music"
          >
            Music
          </option>
          <option
            value="audiobook"
          >
            Audiobook
          </option>
          <option
            value="shortFilm"
          >
            Short Film
          </option>
          <option
            value="tvShow"
          >
            TV Show
          </option>
          <option
            value="software"
          >
            Software
          </option>
          <option
            value="ebook"
          >
            Ebook
          </option>
        </select>
      </div>
      <button
        class="btn btn-primary"
        style="margin-bottom: 10px;"
        type="submit"
      >
        Search
      </button>
    </form>
    <div
      class="result-container"
    />
  </div>
</DocumentFragment>
`;
